### PR TITLE
Defactored Bytes::read

### PR DESCRIPTION
**Pull request by Lucretiel**
Removed unneeded refactoring of read_one_byte, which removed the unneeded dynamic dispatch (`dyn Read`) used by that function.

This function is only used in one place in the entire Rust codebase; there doesn't seem to be a reason for it to exist (and there especially doesn't seem to be a reason for it to use dynamic dispatch)

### CODE REVIEW

**Comment by Lucretiel** [comment]
I did some digging and discovered that the only other place this function was used was in the (now removed) io::Chars iterator (see #27802 and #49970). The dynamic dispatch behavior doesn't appear to have been appropriate back then, either, though the refactoring was; my guess is it was the common mistake of assuming that `&mut Read` is a generic rather than polymorphic construct (huzzah for `dyn` and `impl`)

**Comment by Dylan-DPC** [comment]
ping from triage @shepmaster can you review this?

File: src/libstd/io/mod.rs [hunk 1]
```diff
@@ -1965,7 +1954,15 @@ impl<R: Read> Iterator for Bytes<R> {
...
 
     fn next(&mut self) -> Option<Result<u8>> {
-        read_one_byte(&mut self.inner)
+        let mut byte = 0;
+        loop {
+            return match self.inner.read(slice::from_mut(&mut byte)) {
+                Err(ref e) if e.kind() == ErrorKind::Interrupted => continue,
+                Ok(0) => None,
+                Ok(..) => Some(Ok(byte)),
+                Err(e) => Some(Err(e)),
```
Comment (shepmaster) [inline src/libstd/io/mod.rs]: Why the rearranging of the variants?

Comment (Lucretiel) [reply src/libstd/io/mod.rs]: Force of habit; I like putting my "early escapes" at the top. I can change it back to how it was.

**Comment by shepmaster** [comment]
Thanks!

**Pull request merged**